platform/consul/conversion.go

```diff
@@ -0,0 +1,155 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package consul
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/hashicorp/consul/api"
+	"istio.io/pilot/model"
+)
+
+const (
+	protocolTagName = "protocol"
+	externalTagName = "external"
+)
+
+func convertTags(tags []string) model.Tags {
+	out := make(model.Tags, len(tags))
+	for _, tag := range tags {
+		vals := strings.Split(tag, "=")
+		if len(vals) > 1 {
+			out[vals[0]] = vals[1]
+		} else {
+			// TODO safe to assume all tags are of form "key=value"?
+			out[tag] = ""
+		}
+	}
+	return out
+}
+
+func convertPort(port int, name string) *model.Port {
+
+	// TODO default HTTP/TCP?
+	defaultName := "tcp"
+	if name == "" {
+		name = defaultName
+	}
+
+	return &model.Port{
+		Name:     name,
+		Port:     port,
+		Protocol: convertProtocol(name),
+	}
+}
+
+func convertService(endpoints []*api.CatalogService) *model.Service {
+	name, addr, external := "", "", ""
+
+	node, datacenter := "", ""
+
+	ports := model.PortList{}
+	for _, endpoint := range endpoints {
+		name = endpoint.ServiceName
+		ports = append(ports, convertPort(endpoint.ServicePort, endpoint.NodeMeta[protocolTagName]))
+
+		// TODO This will not work if service is a mix of external and local services
+		// or if a service has more than one external name
+		if endpoint.NodeMeta[externalTagName] != "" {
+			external = endpoint.NodeMeta[externalTagName]
+		}
+
+		// TODO how to handle if a service spans datacenters/nodes?
+		node = endpoint.Node
+		datacenter = endpoint.Datacenter
+
+		// TODO where should the loadbalancer IP be stored in consul?
+		// - Address: IP of consul node for service
+		// - ServiceAddress: IP of service
+		// - TaggedAddresses: list of explicit LAN and WAN IP addresses for the agent
+		// - NodeMetadata
+		addr = endpoint.Address
+	}
+
+	out := &model.Service{
+		Hostname:     serviceHostname(name, node, datacenter),
+		Ports:        ports,
+		Address:      addr,
+		ExternalName: external,
+	}
+
+	return out
+}
+
+func convertInstance(inst *api.CatalogService) *model.ServiceInstance {
+	tags := convertTags(inst.ServiceTags)
+	port := convertPort(inst.ServicePort, inst.NodeMeta[protocolTagName])
+
+	return &model.ServiceInstance{
+		Endpoint: model.NetworkEndpoint{
+			Address:     inst.Address,
+			Port:        inst.ServicePort,
+			ServicePort: port,
+		},
+		Service: &model.Service{
+			Hostname: serviceHostname(inst.ServiceName, "TODO", inst.Datacenter),
+			Address:  inst.Address,
+			Ports:    model.PortList{port},
+			// TODO ExternalName come from metadata?
+			ExternalName: inst.NodeMeta[externalTagName],
+		},
+		Tags: tags,
+	}
+}
+
+// serviceHostname produces FQDN for a consul service
+func serviceHostname(name, node, datacenter string) string {
+	// TODO include consul node in Hostname?
```

Comment:
at pilot won't we need to know all the nodes that make up a service? For example, reviews (rev. v1, v2, v3) and their node ip's inside k8s would be known?
otherwise how will a micro service hosted in consul be able to apply a destination based policy?